### ISSUE TITLE
Failed Solr integration tests caused by SecureRandom algorithm "NativePRNG"

### ISSUE THREAD

**Issue by adamyy** [castorini/anserini#1367]
Environment:

- openjdk 11.0.7 2020-04-14
- OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.7+10)
- OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.7+10, mixed mode)

I have observed the following test failures in my local build of master:

```
java.lang.AssertionError: SecureRandom algorithm 'NativePRNG' is in use by your JVM, which is a potentially blocking algorithm on some environments. Please report the details of this failure (and your JVM vendor/version) to solr-user@lucene.apache.org. Y
ou can try to run your tests with -Djava.security.egd=file:/dev/./urandom or bypass this check using -Dtest.solr.allowed.securerandom=NativePRNG as a JVM option when running tests.
```

**TL;DR** if you do experience this issue, please try again using `-Djava.security.egd=file:/dev/./urandom`.

It seems like the Solr test framework is forbidding the use of certain SecureRandom algorithm, namely `NativePRNG` and `NativePRNGBlocking`. The related Solr [thread](https://issues.apache.org/jira/browse/SOLR-10338) suggested that `SHA1PRNG` should be preferred in tests for its nonblocking nature. 

I suspect that this issue is JDK dependent, for instance the TravisCI [build](https://travis-ci.org/github/castorini/anserini/jobs/720180310) wasn't impacted. 

There is a number of ways to alleviate this issue if you do experience it. We can set the system property `java.security.egd` to `file:/dev/./urandom` for all tests, which change the default SecureRandom algorithm to `SHA1PRNG`. Ideally I think the configuration should be set only for Solr tests, still trying to figure out how to do that.

To set the system property for all tests, either supply the option in the command line, or set that in the Maven surefire plugin:

```xml
<plugin>
  <groupId>org.apache.maven.plugins</groupId>
  <artifactId>maven-surefire-plugin</artifactId>
  <version>2.12.4</version>
  <configuration>
    <systemPropertyVariables>
      <java.security.egd>file:/dev/./urandom</java.security.egd>
    </systemPropertyVariables>
  </configuration>
</plugin>
```

Full logs:

```
Running io.anserini.integration.solr.TrecEndToEndTest
2020-09-05 17:34:46,382 WARN  [SUITE-TrecEndToEndTest-seed#[2B632D1E983E89A3]-worker] solr.SolrTestCaseJ4 (SolrTestCaseJ4.java:2879) - System property java.security.egd was not set by test runner, forcibly set to expected: file:/dev/./urandom
NOTE: test params are: codec=Asserting(Lucene80), sim=Asserting(org.apache.lucene.search.similarities.AssertingSimilarity@365e10a6), locale=sl, timezone=Asia/Dili
NOTE: Mac OS X 10.15.5 x86_64/AdoptOpenJDK 11.0.7 (64-bit)/cpus=16,threads=1,free=280314808,total=310378496
NOTE: All tests run in this JVM: [TwentyNewsgroupsCollectionTest, FeverSentenceCollectionTest, JsonCollectionDocumentArrayTest, TrecCollectionTest, Cord19ParagraphCollectionTest, CoreCollectionTest, TweetCollectionTest, CommonCrawlWarcCollectionTest, Cor
d19AbstractCollectionTest, Iso19115CollectionTest, WashingtonPostCollectionTest, WikipediaCollectionTest, JsonCollectionDocumentObjectTest, NewYorkTimesCollectionTest, BibtexCollectionTest, Cord19FullTextCollectionTest, CleanTrecCollectionTest, TrecwebCo
llectionTest, ClueWeb12CollectionTest, CommonCrawlNewsEnWarcCollectionTest, FeverParagraphCollectionTest, ClueWeb09CollectionTest, JsonCollectionLineObjectTest, AclAnthologyTest, CommonCrawlWetCollectionTest, ExtractTopDfTermsTest, ExtractNormsTest, Feat
ureVectorTest, ExtractAverageDocumentLengthTest, ExtractDocumentLengthsTest, TweetEndToEndTest, BibtexEndToEndTest, MultiThreadingSearchTest, AclAnthologyEndToEndTest, CoreEndToEndTest, AclAnthologyEndToEndTest, CoreEndToEndTest, TrecEndToEndTest]
NOTE: reproduce with: ant test  -Dtestcase=TrecEndToEndTest -Dtests.seed=2B632D1E983E89A3 -Dtests.slow=true -Dtests.badapples=true -Dtests.locale=sl -Dtests.timezone=Asia/Dili -Dtests.asserts=true -Dtests.file.encoding=UTF-8
Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.261 sec <<< FAILURE!
io.anserini.integration.solr.TrecEndToEndTest  Time elapsed: 0.259 sec  <<< FAILURE!
java.lang.AssertionError: SecureRandom algorithm 'NativePRNG' is in use by your JVM, which is a potentially blocking algorithm on some environments. Please report the details of this failure (and your JVM vendor/version) to solr-user@lucene.apache.org. Y
ou can try to run your tests with -Djava.security.egd=file:/dev/./urandom or bypass this check using -Dtest.solr.allowed.securerandom=NativePRNG as a JVM option when running tests.
        at __randomizedtesting.SeedInfo.seed([2B632D1E983E89A3]:0)
        at org.junit.Assert.fail(Assert.java:88)
        at org.junit.Assert.assertTrue(Assert.java:41)
        at org.junit.Assert.assertFalse(Assert.java:64)
        at org.apache.solr.SolrTestCaseJ4.assertNonBlockingRandomGeneratorAvailable(SolrTestCaseJ4.java:2886)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1750)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:878)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:894)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at com.carrotsearch.randomizedtesting.rules.SystemPropertiesRestoreRule$1.evaluate(SystemPropertiesRestoreRule.java:57)
        at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at org.apache.lucene.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:41)
        at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
        at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
        at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
        at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
        at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:54)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
        at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by adamyy** [comment]
Was going to assign this to myself, but I don't seem to have the permission 😅 

My apology for causing the headache!

**Comment by adamyy** [comment]
Also thinking about the way I wrote the integration tests. 

There is only a handful of docs that I can find regarding how to write integration tests with Solr. Many of them points me to the embedded Solr server, and the `EmbeddedSolrServerTestBase` class included in the Solr-test-framework was very tempting as it seems to handle many of the heavy lifting around creating/destroying the embedded Solr server. 

However, I can sort of smell the developer's reluctance of exposing this class to the end users, from the lack of documentation and the response to [this user's comment](https://issues.apache.org/jira/browse/SOLR-10338) regarding using this class in an external project.

Currently experimenting with bypassing the all-powerful `EmbeddedSolrServerTestBase` and just use the `EmbeddedSolrServer` on its own. If this is possible, we should have no more worry about this RNG algorithm assertion.

**Comment by lintool** [comment]
Okay, thanks for the careful thought put into this!

**Comment by lintool** [comment]
@adamyy I think this issue can be closed?

**Comment by adamyy** [comment]
@lintool I believe the issue was resolved! I tested locally with `mvn surefire:test` and `mvn clean package appassembler:assemble`.
